### PR TITLE
Closes #21

### LINKED ISSUE #21: Create JWT issuer
Add JWT and refresh token issuer

### PR DESCRIPTION
**Pull request by mtracewicz**
# This commit closes isses: #21, #20, #18, #17

## Description
I have implemented user accounts based on email and password.
Moreover, registration and login have been created with JWT.
For this, there are three endpoints with the prefix `/api/auth/`.
The endpoints are:

- `login`
- `register`
- `refresh-token`

For this, I needed a Postgre instance so I have added it to docker-compose. As this is a development one I exposed the port to the public so We can connect to DB and check it.

## How to test
Run this with docker-compose - first, check the updated README (You should not need to set the variables with JWT prefix as they should load from `appsetings.development.json`)

First, register user using `/api/auth/register`  (also here should be tested if the user can't be created twice with the same email, password rules are applied, etc.)

Second login using credentials from the previous step `/api/auth/login`

Using the token from this step try to access `/api/greetings/user`  (try if it can be accessed only with this token
Then hit the endpoint  `/api/greetings/helloworld` (this one should be accessible with and without the token as this is a public endpoint)

Afterward, try to request a new token using the refresh token `/api/auth/refresh-token` check if the newly obtained JWT works as well.

For the last step login to the database and set `IsRevoked` for your refresh token to true. Wait 15 minutes (the time till which JWT is valid) and try to request a new token with `/api/auth/refresh-token` . This should now fail.

Lastly use the `/api/auth/logout` to log out. This requires authentication with the current JWT.
This endpoint deletes all currently working refresh tokens for the user. This DOES NOT invalidate current JWT as this is impossible so users can still use it till it is expired.

### CODE REVIEW

File: backend/src/api/Controllers/AuthenticationController.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        else
+        {
+            return BadRequest(Error);
+        }
+    }
+
+    [HttpPost("login")]
+    [ProducesResponseType(200)]
+    [ProducesResponseType(401)]
```
Comment (Sojusan) [inline backend/src/api/Controllers/AuthenticationController.cs:40]: Shouldn't there be also `[ProducesResponseType(400)]` as there is a possibility of `BadRequest` (no login credentials)?

File: backend/src/api/Controllers/GreetingsController.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+namespace api.Controllers;
+
+[ApiController]
+[Route("/api/greetings")]
+public class GreetingsController : ControllerBase
+{
+    [HttpGet]
+    [Authorize]
+    [Route("user")]
+    public IActionResult Greet()
```
Comment (Sojusan) [inline backend/src/api/Controllers/GreetingsController.cs:13]: What about `[ProducesResponseType(200)]`? It is not necessary to add?

File: backend/src/api/Services/Users/UserService.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            var (authresult, refreshToken) = _tokenService.Generate(existingUser);
+            await _usersDbContext.RefreshTokens.AddAsync(refreshToken);
+            await _usersDbContext.SaveChangesAsync();
+            return (true, authresult, null);
+        }
+        else
+        {
+            return (false, null, "Invalid password");
+        }
+    }
```
Comment (Sojusan) [inline backend/src/api/Services/Users/UserService.cs:71]:
```suggestion
    }

```

File: backend/src/api/Services/Users/UserService.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public async Task<(bool IsSuccess, AuthResultDTO AuthResult, string Error)> RefreshLogin(TokenRequestDTO tokenRequestDTO)
+    {
+        var storedToken = _usersDbContext.RefreshTokens.FirstOrDefault(t => t.Token.Equals(tokenRequestDTO.RefreshToken));
+        if (storedToken == null)
+        {
+            return (false, null, "Refresh token is invalid!")
+;
```
Comment (Sojusan) [inline backend/src/api/Services/Users/UserService.cs:78]: The semicolon at the beginning of the new line?

**Review by Sojusan — CHANGES_REQUESTED** [review]
I got some questions and ideas.

File: backend/src/api/Controllers/AuthenticationController.cs (hunk 1 above)
Comment (mtracewicz) [reply backend/src/api/Controllers/AuthenticationController.cs:40]: It is only used to creat a Swagger document not necessary, good to add

File: README.md [hunk 5]
```diff
@@ -3,6 +3,37 @@
...
+
+- docker
+- docker-compose
+- powershell
+- dotnet sdk
+- entity framework tools
+
+First set up the following environment variables:
+
+- JWT\_\_SECRET
```
[the remark this answers is not in the document]
Comment (mtracewicz) [reply README.md:18]: Double underscore maps to the following appsettings:
```
{
   "JWT":{
    "SECRET" : "VA:UE"
   }
}
```

File: README.md [hunk 6]
```diff
@@ -3,6 +3,37 @@
...
+
+To run this application locally it is recommended to have the following installed:
+
+- docker
+- docker-compose
+- powershell
+- dotnet sdk
+- entity framework tools
+
+First set up the following environment variables:
```
[the remark this answers is not in the document]
Comment (mtracewicz) [reply README.md:16]: Might be a good idea for a later time :)

File: scripts/apply_migrations.ps1 [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+docker run --network host -e $passwd -v ${src}:/migrations/ --rm postgres `
+    psql -h localhost -U $Env:POSTGRES_USER -d users -f /migrations/migration.sql `
+    2>&1 > $null
+
+Write-Output "Cleanup"
+Remove-Item -Recurse -Force migration_scripts
+Set-Location ../../../scripts
+Write-Output "Migrations had been applyed!"
+Write-Host "Stoping DB"
+docker stop ksummarizedcom-db-1 2>&1 > $null
```
[the remark this answers is not in the document]
Comment (mtracewicz) [reply scripts/apply_migrations.ps1:20]: Feel free to provide a change :)

File: README.md [hunk 8]
```diff
@@ -3,6 +3,37 @@
...
+
+- JWT\_\_SECRET
+- JWT\_\_AUDIENCE
+- JWT\_\_ISSUER
+- POSTGRES_PASSWORD
+- POSTGRES_USER
+
+Next go to the `scripts` directory and run `apply_migrations.ps1`
```
[the remark this answers is not in the document]
Comment (mtracewicz) [reply README.md:24]: Yeah might be a good change

File: backend/src/api/Services/Users/UserService.cs (hunk 3 above)
Comment (mtracewicz) [reply backend/src/api/Services/Users/UserService.cs:71]: Will update in a future PR

File: backend/src/api/Services/Users/UserService.cs (hunk 4 above)
Comment (mtracewicz) [reply backend/src/api/Services/Users/UserService.cs:78]: A formatting error should be corrected in next PR

**Review by Sojusan — APPROVED** [review]

**Pull request merged**